File: express-service/src/routes/parse.ts
/**
 * Route for LLM-powered address parsing and validation endpoint.
 *
 * Exposes POST /validate/parse which accepts raw unstructured address text,
 * parses it via the LLM parser into structured fields, then validates the
 * parsed address via USPS.
 */

import { Router, Request, Response } from 'express';
import { AddressValidator } from '../services/addressValidator';
import { LLMParser, LLMTimeoutError, LLMParseError } from '../parsers/llmParser';
import { ParseRequest, ParseValidationResult } from '../models/validation';

/**
 * Create the parse router with injected dependencies.
 *
 * @param addressValidator - The address validation service
 * @param llmParser - The LLM-powered address parser
 * @returns Express Router handling POST /validate/parse
 */
export function createParseRouter(addressValidator: AddressValidator, llmParser: LLMParser): Router {
  const router = Router();

  router.post('/validate/parse', async (req: Request, res: Response): Promise<void> => {
    const body = req.body as Partial<ParseRequest>;

    // Validate request body
    if (!body.raw_address || typeof body.raw_address !== 'string') {
      res.status(400).json({
        detail: {
          error_code: 'VALIDATION_ERROR',
          message: 'raw_address is required and must be a string',
        },
      });
      return;
    }

    const rawText = body.raw_address;

    // Step 1: Parse raw text into structured address via LLM
    try {
      const parsedAddress = await llmParser.parse(rawText);

      // Step 2: Validate the parsed address via USPS
      const validationResult = await addressValidator.validate(parsedAddress);

      // Determine overall status based on validation result
      const status = validationResult.status === 'valid' || validationResult.status === 'invalid'
        ? validationResult.status
        : 'invalid';

      const result: ParseValidationResult = {
        raw_text: rawText,
        parsed_address: parsedAddress,
        validation_result: validationResult,
        status,
        error_message: validationResult.error_message,
      };

      res.status(200).json(result);
    } catch (error: unknown) {
      if (error instanceof LLMTimeoutError) {
        const result: ParseValidationResult = {
          raw_text: rawText,
          parsed_address: undefined,
          validation_result: undefined,
          status: 'service_unavailable',
          error_message: 'LLM service is unavailable',
        };
        res.status(200).json(result);
        return;
      }

      if (error instanceof LLMParseError) {
        const result: ParseValidationResult = {
          raw_text: rawText,
          parsed_address: undefined,
          validation_result: undefined,
          status: 'parse_failed',
          error_message: error.message,
        };
        res.status(200).json(result);
        return;
      }

      // Unexpected error
      res.status(500).json({
        detail: {
          error_code: 'INTERNAL_ERROR',
          message: 'An unexpected error occurred',
        },
      });
    }
  });

  return router;
}

File: python-service/app/routes/parse.py
"""Route for LLM-powered address parsing and validation endpoint.

Exposes POST /validate/parse which accepts raw unstructured address text,
parses it via the LLM parser into structured fields, then validates the
parsed address via USPS.
"""

from fastapi import APIRouter, Depends

from app.adapters.usps_adapter import USPSAdapterInterface, get_usps_adapter
from app.models.validation import ParseRequest, ParseValidationResult
from app.parsers.llm_parser import LLMParser, LLMParseError, LLMTimeoutError
from app.services.address_validator import AddressValidator

router = APIRouter()


def _get_address_validator(
    adapter: USPSAdapterInterface = Depends(get_usps_adapter),
) -> AddressValidator:
    """Dependency that provides an AddressValidator wired with the USPS adapter."""
    return AddressValidator(usps_adapter=adapter)


def _get_llm_parser() -> LLMParser:
    """Dependency that provides an LLMParser instance."""
    return LLMParser()


@router.post("/validate/parse", response_model=ParseValidationResult)
async def parse_and_validate(
    request: ParseRequest,
    validator: AddressValidator = Depends(_get_address_validator),
    parser: LLMParser = Depends(_get_llm_parser),
) -> ParseValidationResult:
    """Parse unstructured address text via LLM and validate via USPS.

    Accepts a ParseRequest with a raw_address string. The LLM parser attempts
    to extract structured address fields from the text. If successful, the
    parsed address is validated against USPS.

    Error scenarios return HTTP 200 with an appropriate status in the body:
    - LLMTimeoutError → status "service_unavailable"
    - LLMParseError → status "parse_failed"
    """
    raw_text = request.raw_address

    # Step 1: Parse raw text into structured address via LLM
    try:
        parsed_address = await parser.parse(raw_text)
    except LLMTimeoutError:
        return ParseValidationResult(
            raw_text=raw_text,
            parsed_address=None,
            validation_result=None,
            status="service_unavailable",
            error_message="LLM service is unavailable",
        )
    except LLMParseError as exc:
        return ParseValidationResult(
            raw_text=raw_text,
            parsed_address=None,
            validation_result=None,
            status="parse_failed",
            error_message=str(exc),
        )
    except (NotImplementedError, Exception):
        return ParseValidationResult(
            raw_text=raw_text,
            parsed_address=None,
            validation_result=None,
            status="service_unavailable",
            error_message="LLM service is unavailable",
        )

    # Step 2: Validate the parsed address via USPS
    validation_result = await validator.validate(parsed_address)

    # Determine overall status based on validation result
    status = validation_result.status if validation_result.status in ("valid", "invalid") else "invalid"

    return ParseValidationResult(
        raw_text=raw_text,
        parsed_address=parsed_address,
        validation_result=validation_result,
        status=status,
        error_message=validation_result.error_message,
    )
